google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkWriter.java

```diff
@@ -661,8 +661,10 @@ private ApiFuture<WriteResult> executeWrite(
         ApiFutures.transformAsync(
             operation.getFuture(),
             result -> {
-              pendingOpsCount--;
-              processBufferedOperations();
+              synchronized (lock) {
+                pendingOpsCount--;
+                processBufferedOperations();
+              }
```

Comment:
Synchronization block to ensure that buffered operations are processed in expected order